tests/stability/test_adaptive_scaling.py

```diff
@@ -0,0 +1,302 @@
+    """Tests that adaptive scaling reacts in a reasonable amount of time to
+    an increased task load and scales up.
+    """
+    maximum = 10
+    with Cluster(
+        name=f"test_adaptive_scaling-{uuid.uuid4().hex}",
+        n_workers=minimum,
+        worker_vm_types=["t3.medium"],
+        wait_for_workers=True,
+        # Note: We set allowed-failures to ensure that no tasks are not retried upon ungraceful shutdown behavior
+        # during adaptive scaling but we receive a KilledWorker() instead.
+        environ={"DASK_DISTRIBUTED__SCHEDULER__ALLOWED_FAILURES": "0"},
+    ) as cluster:
+        with Client(cluster) as client:
+            adapt = cluster.adapt(minimum=minimum, maximum=maximum)
+            time.sleep(adapt.interval * 2.1)  # Ensure enough time for system to adapt
+            assert len(adapt.log) == 0
+            ev_fan_out = Event(name="fan-out", client=client)
+
+            def clog(x: int, ev: Event) -> int:
+                ev.wait()
+                return x
+
+            futures = client.map(clog, range(100), ev=ev_fan_out)
+
+            start = time.monotonic()
+            client.wait_for_workers(n_workers=maximum, timeout=TIMEOUT_THRESHOLD)
+            end = time.monotonic()
+            duration = end - start
+            assert duration < threshold, duration
+            assert len(adapt.log) <= 2
+            assert adapt.log[-1][1] == {"status": "up", "n": maximum}
+            ev_fan_out.set()
+            client.gather(futures)
+            return duration
+
+
+@pytest.mark.skip(reason="coiled-runtime#266")
+@pytest.mark.stability
+def test_adapt_to_changing_workload():
+    """Tests that adaptive scaling reacts within a reasonable amount of time to
+    a varying task load and scales up or down. This also asserts that no recomputation
+    is caused by the scaling.
+    """
+    minimum = 0
+    maximum = 10
+    fan_out_size = 100
+    with Cluster(
+        name=f"test_adaptive_scaling-{uuid.uuid4().hex}",
+        n_workers=maximum,
+        worker_vm_types=["t3.medium"],
+        wait_for_workers=True,
+        # Note: We set allowed-failures to ensure that no tasks are not retried upon ungraceful shutdown behavior
+        # during adaptive scaling but we receive a KilledWorker() instead.
+        environ={"DASK_DISTRIBUTED__SCHEDULER__ALLOWED_FAILURES": "0"},
+    ) as cluster:
+        with Client(cluster) as client:
+
+            @delayed
+            def clog(x: int, ev: Event, **kwargs) -> int:
+                ev.wait()
+                return x
+
+            def workload(
+                fan_out_size,
+                ev_fan_out,
+                ev_barrier,
+                ev_final_fan_out,
+            ):
+                fan_out = [clog(i, ev=ev_fan_out) for i in range(fan_out_size)]
+                barrier = clog(delayed(sum)(fan_out), ev=ev_barrier)
+                final_fan_out = [
+                    clog(i, ev=ev_final_fan_out, barrier=barrier)
```

Comment:
What's the barrier used for here?

Reply:
The idea is to start the `final_fan_out` tasks once the `barrier` task, i.e., scaling down, is done. Would replacing `barrier` with `reduction` make this clearer?